### PR TITLE
support Megatron GRPO with vLLM on AMD

### PR DESCRIPTION
**Pull request by amd-pratmish**
## Summary

ROCm path for end-to-end GRPO using SkyRL's existing Megatron backend and vLLM rollout engine. No separate training backend or vLLM fork is introduced.

HIP-specific defaults are gated so existing behavior stays unchanged.

## Validation

2-GPU colocated GSM8K recipe (`Qwen/Qwen2.5-0.5B-Instruct`), covering vLLM generation, colocated weight synchronization, and Megatron policy updates.

| GPU | Architecture | ISA | Status |
| --- | --- | --- | --- |
| MI300X | CDNA3 | gfx942 | Supported |
| MI325X | CDNA3 | gfx942 | Supported |
| MI355X | CDNA4 | gfx950 | Validated |

## What changes

- Adds `integrations/rocm_amd/` for GPU detection, installation, vLLM build/cache, preflight, smoke tests, and reviewer reproduction.
- Adds a 2-GPU colocated GSM8K Megatron GRPO recipe.
- Aligns HIP/ROCR/CUDA device masks for Ray-managed vLLM workers (physical ROCr IDs, logical HIP/CUDA IDs) and applies them only on HIP.
- Pins GPU visibility in the actor `runtime_env` and before vLLM import on colocated `mp` engines.
- Sets ROCm vLLM worker env (`VLLM_USE_V1=0`, spawn) before engine init.
- Caps Ray object-store memory on HIP so colocated vLLM is not starved.
- Adds compatibility shims for the tested vLLM OpenAI serving APIs and vLLM-router PD flags.
- Honors `data.dataloader.num_workers` in prompt-dataset construction (defaults remain 8).

## Reproduce

Requirements: Docker or Podman, `/dev/kfd`, `/dev/dri`, and two supported GPUs.

```bash
ROCM_IMAGE=rocm/primus:v26.4 ROCM_GPUS=2 \
  bash integrations/rocm_amd/run_in_container.sh

# Inside the container
bash integrations/rocm_amd/install_full_stack.sh
python3 integrations/rocm_amd/gpu_support.py
NUM_GPUS=2 bash examples/train/gsm8k/run_gsm8k_megatron_rocm.sh
```

Full instructions and pass criteria: `integrations/rocm_amd/REPRODUCE.md`.

## Pinned dependencies

| Dependency | Pin |
| --- | --- |
| Base image | [rocm/primus:v26.4](https://hub.docker.com/r/rocm/primus) |
| megatron-core | [71e418ea](https://github.com/NVIDIA/Megatron-LM/commit/71e418ea7d7b3a6c9a53238c543c3e0b43e11026) |
| Megatron-Bridge | [91a15142](https://github.com/NVIDIA-NeMo/Megatron-Bridge/commit/91a15142a4b4442a8d46ab539d1b923bd08570d0) |
| vLLM | [bc150f50](https://github.com/vllm-project/vllm/commit/bc150f50299199599673614f80d12a196f377655) (`v0.20.2`), built against the container's ROCm PyTorch |
| Ray | 2.57.0 |
| Model | [Qwen/Qwen2.5-0.5B-Instruct](https://huggingface.co/Qwen/Qwen2.5-0.5B-Instruct) |
| Dataset | [openai/gsm8k](https://huggingface.co/datasets/openai/gsm8k) |

## Test plan

- [ ] GPU allowlist / ISA detection unit tests
- [ ] `NUM_GPUS=2 bash examples/train/gsm8k/run_gsm8k_megatron_rocm.sh` on Instinct (MI300X / MI325X / MI355X)
- [ ] Confirm CUDA/non-HIP defaults are unchanged

## Verification

- GPU allowlist and ISA detection unit tests
- MI355X 2-GPU Megatron GRPO (`Training done!`, exit 0)

---

> [!NOTE]
> **Medium Risk**
> Touches Ray placement, vLLM serving, and Megatron weight sync on a new platform; changes are mostly HIP-gated but inference validation and pinned dependency drift remain operational risks.
> 
> **Overview**
> Adds **AMD Instinct (ROCm) support** for colocated **Megatron GRPO + local vLLM rollout** on the existing `trainer.strategy=megatron` path—no new training backend.
> 
> **Integration & ops:** New `integrations/rocm_amd/` (GPU allowlist, pinned megatron-core/Bridge/vLLM install & wheel cache, container/Slurm launchers, preflight/smoke/matrix scripts) plus `docker/Dockerfile.rocm-megatron`, `docker/pyproject.rocm.toml`, and `examples/train/gsm8k/run_gsm8k_megatron_rocm.sh`. Root README links to the integration docs.
> 
> **Runtime (HIP-gated):** Aligns **ROCr physical vs HIP/CUDA logical** device masks for colocated `num_gpus=0` vLLM actors; sets `RAY_ACCEL_ENV_VAR_OVERRIDE_ON_ZERO=0`, defers vLLM imports until masks are applied, and injects ROCm vLLM defaults (`VLLM_USE_V1=0`, `VLLM_TARGET_DEVICE=rocm`, spawn). On ROCm, colocated single-node vLLM auto-switches **`distributed_executor_backend` from `ray` to `mp`**; Ray init gets explicit `num_gpus` and a smaller object store. Megatron workers gain **`bridge_compat`** (`load_auto_bridge`, LoRA import fallbacks) for pinned Bridge revisions.
> 
> **Small cross-platform tweaks:** Prompt datasets use `data.dataloader.num_workers`; vLLM server/router tolerate pinned **v0.20.2** API differences (`create_server_socket`, PD router flags).
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit b16d6da4a98dad6f2ebd997a63a2d673f2c96e4f. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

### CODE REVIEW

File: examples/train/gsm8k/run_gsm8k_megatron_rocm.sh [hunk 1]
```diff
@@ -0,0 +1,108 @@
...
+export VLLM_USE_V1=0
+export VLLM_USE_TRITON_FLASH_ATTN=0
+export FLA_TILELANG=0
+# Keep PyTorch's CUDA-compatible device mask aligned with ROCm masks.
+if [[ -n "${HIP_VISIBLE_DEVICES:-}" ]]; then
+  export CUDA_VISIBLE_DEVICES="${HIP_VISIBLE_DEVICES}"
+  export ROCR_VISIBLE_DEVICES="${HIP_VISIBLE_DEVICES}"
+else
+  unset CUDA_VISIBLE_DEVICES
+fi
```
[the remark this answers is not in the document]
Comment (amd-pratmish) [reply examples/train/gsm8k/run_gsm8k_megatron_rocm.sh:61]: Fixed in e34783d0. The recipe now preserves physical IDs in ROCR_VISIBLE_DEVICES and remaps HIP/CUDA to process-local 0..N-1 indices. The full 2-GPU MI355X E2E passed after this change.

File: skyrl/train/utils/utils.py [hunk 2]
```diff
@@ -654,6 +654,26 @@ def validate_inference_engine_cfg(cfg: SkyRLTrainConfig):
...
+        ie_cfg.distributed_executor_backend == "ray"
+        and cfg.trainer.placement.colocate_all
+        and ie_cfg.backend == "vllm"
+    ):
+        try:
+            import torch as _torch_ie
+
+            _on_rocm_ie = getattr(_torch_ie.version, "hip", None) is not None
+        except ImportError:
+            _on_rocm_ie = False
```
[the remark this answers is not in the document]
Comment (amd-pratmish) [reply skyrl/train/utils/utils.py]: Addressed in e34783d0 by using the existing module-level torch import and direct HIP checks.

File: integrations/rocm_amd/install_vllm_rocm.sh [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+  echo "Installing cached wheel ${wheel}"
+  python3 -m pip install --no-cache-dir --force-reinstall --no-deps "${wheel}"
+  install_runtime_deps
+  assert_torch_unchanged
+  python3 "${VERIFY_PY}" || return 1
+}
+
+WHEEL=""
+if compgen -G "${WHEEL_CACHE}/vllm-0.20*.whl" >/dev/null; then
+  WHEEL="$(ls -1t "${WHEEL_CACHE}"/vllm-0.20*.whl | head -1)"
```
[the remark this answers is not in the document]
Comment (amd-pratmish) [reply integrations/rocm_amd/install_vllm_rocm.sh:47]: Fixed in e34783d0. Cached wheels are now partitioned by vLLM revision, exact torch and HIP versions, Python SOABI, and PYTORCH_ROCM_ARCH. A fresh ABI-keyed wheel was rebuilt and passed the full 2-GPU E2E.

File: examples/train/gsm8k/run_gsm8k_megatron_rocm.sh [hunk 4]
```diff
@@ -0,0 +1,108 @@
...
+: "${CKPT_DIR:=$HOME/ckpts/gsm8k_megatron_rocm}"
+
+mkdir -p "$DATA_DIR" "$CKPT_DIR" "$LOG_DIR"
+
+echo "=== Ray + vLLM preflight ==="
+bash integrations/rocm_amd/ray_preflight.sh
+python3 integrations/rocm_amd/verify_vllm_skyrl_compat.py
+
+echo "=== Preparing GSM8K (tiny subset) ==="
+python3 examples/train/gsm8k/gsm8k_dataset.py \
```
[the remark this answers is not in the document]
Comment (amd-pratmish) [reply examples/train/gsm8k/run_gsm8k_megatron_rocm.sh:35]: The repository rule is followed for development and test commands. These scripts are container bootstrap tooling: they must install Megatron and source-built vLLM into the active base-image environment so the native extensions link against that image’s ROCm PyTorch. Running bootstrap under `uv run --isolated` creates a second environment/PyTorch and breaks that ABI guarantee. I documented this boundary in e34783d0. The resulting image passed the complete 2-GPU E2E.

File: skyrl/train/utils/utils.py [hunk 5]
```diff
@@ -654,6 +654,26 @@ def validate_inference_engine_cfg(cfg: SkyRLTrainConfig):
...
+        if _on_rocm_ie:
+            logger.info(
+                "ROCm colocated vLLM: switching distributed_executor_backend from ray to mp "
+                "(vLLM Ray executor is incompatible with ROCm GPU visibility in colocated GRPO)."
+            )
+            ie_cfg.distributed_executor_backend = "mp"
```
[the remark this answers is not in the document]
Comment (amd-pratmish) [reply skyrl/train/utils/utils.py]: Fixed in e34783d0. The ROCm ray-to-mp fallback now applies only when TP*PP fits on one node; cross-node engines retain the requested Ray backend. Added a regression test.

File: skyrl/train/utils/utils.py [hunk 6]
```diff
@@ -811,11 +831,28 @@ def prepare_runtime_environment(cfg: SkyRLTrainConfig) -> dict[str, str]:
...
-        if cfg.trainer.flash_attn:
+        # CUDA-only: disabling fused TE attention. On ROCm/HIP this breaks Megatron-Bridge
+        # model materialize (NVTE_FLASH_ATTN assertion). Skip on AMD.
+        _on_rocm = False
+        try:
+            import torch
+
+            _on_rocm = getattr(torch.version, "hip", None) is not None
+        except ImportError:
+            pass
```
[the remark this answers is not in the document]
Comment (amd-pratmish) [reply skyrl/train/utils/utils.py]: Addressed in e34783d0 by using the existing module-level torch import and direct HIP checks.

File: docker/Dockerfile.rocm-megatron [hunk 7]
```diff
@@ -0,0 +1,65 @@
...
+        iproute2 \
+        kmod \
+        libnuma-dev \
+        libxml2 \
+        netcat-openbsd \
+        tzdata \
+    && apt-get clean \
+    && rm -rf /var/lib/apt/lists/*
+
+RUN if ! command -v uv >/dev/null 2>&1; then \
```
[the remark this answers is not in the document]
Comment (amd-pratmish) [reply docker/Dockerfile.rocm-megatron:47]: Fixed in e34783d0. The uv installer is version-pinned, downloaded to disk, and verified against a pinned SHA-256 before execution.

File: skyrl/train/utils/utils.py [hunk 8]
```diff
@@ -824,6 +861,15 @@ def prepare_runtime_environment(cfg: SkyRLTrainConfig) -> dict[str, str]:
...
                 env_vars[nvte_var] = os.environ[nvte_var]
 
+    try:
+        import torch as _torch
+
+        if getattr(_torch.version, "hip", None) is not None:
+            env_vars.setdefault("VLLM_TARGET_DEVICE", "rocm")
+            env_vars.setdefault("VLLM_WORKER_MULTIPROC_METHOD", "spawn")
+    except ImportError:
+        pass
```
[the remark this answers is not in the document]
Comment (amd-pratmish) [reply skyrl/train/utils/utils.py]: Addressed in e34783d0 by using the existing module-level torch import and direct HIP checks.

File: skyrl/train/utils/utils.py [hunk 9]
```diff
@@ -842,12 +888,29 @@ def prepare_runtime_environment(cfg: SkyRLTrainConfig) -> dict[str, str]:
...
+        if configured_vllm_use_v1 is not None:
+            env_vars["VLLM_USE_V1"] = configured_vllm_use_v1
+        else:
+            _on_rocm_vllm = False
+            try:
+                import torch as _torch_vllm
+
+                _on_rocm_vllm = getattr(_torch_vllm.version, "hip", None) is not None
+            except ImportError:
+                pass
```
[the remark this answers is not in the document]
Comment (amd-pratmish) [reply skyrl/train/utils/utils.py]: Addressed in e34783d0 by using the existing module-level torch import and direct HIP checks.

File: skyrl/train/utils/utils.py [hunk 10]
```diff
@@ -1111,7 +1191,24 @@ def initialize_ray(cfg: SkyRLTrainConfig):
...
+        "runtime_env": {"env_vars": env_vars},
+        "log_to_driver": True,
+    }
+    _on_rocm_ray = False
+    try:
+        import torch as _torch_ray
+
+        _on_rocm_ray = getattr(_torch_ray.version, "hip", None) is not None
+    except ImportError:
+        pass
```
[the remark this answers is not in the document]
Comment (amd-pratmish) [reply skyrl/train/utils/utils.py]: Addressed in e34783d0 by using the existing module-level torch import and direct HIP checks.

File: skyrl/backends/skyrl_train/inference_servers/vllm_server_actor.py [hunk 11]
```diff
@@ -246,7 +242,7 @@ def __init__(
 
         # Configure GPU visibility for this server's TP/PP workers
         if self._use_mp_backend:
-            self._setup_mp_gpu_visibility(mp_cuda_visible_devices)
+            pass  # applied at start of __init__ before vLLM imports
         else:
             os.environ["VLLM_RAY_PER_WORKER_GPUS"] = str(0.2 if colocated_training else 1.0)
             # Set bundle indices for this server's TP/PP workers in the placement group.
```
[the remark this answers is not in the document]
Comment (amd-pratmish) [reply skyrl/backends/skyrl_train/inference_servers/vllm_server_actor.py]: Addressed in e34783d0 by inverting the condition and removing the empty branch.

**Comment by amd-pratmish** [comment]
Addressed the review feedback in e34783d0 and 4358d887:

- corrected physical ROCr vs logical HIP/CUDA device masks
- retained Ray for cross-node ROCm vLLM engines and added a regression test
- keyed cached vLLM wheels by revision, torch/HIP versions, Python ABI, and target ISAs
- verified the pinned uv installer SHA-256 before execution
- applied the requested torch-import and conditional cleanup
- merged the latest upstream `main`

Validation on 2x MI355X completed successfully after rebuilding vLLM through the new cache path: vLLM/SkyRL compatibility passed, Megatron GRPO reached `Training done!`, and the workflow ended with `ALL VERIFY CHECKS PASSED`.

**Comment by amd-pratmish** [comment]
> Cursor Bugbot has reviewed your changes using default effort and found 1 potential issue.
> 
> There are 6 total unresolved issues (including 5 from previous reviews).
> 
> [Fix All in Cursor](https://cursor.com/open?link=eyJ2ZXJzaW9uIjoxLCJ0eXBlIjoiQlVHQk9UX0ZJWF9BTExfSU5fQ1VSU09SIiwiZGF0YSI6eyJyZWRpc0tleSI6ImJ1Z2JvdC1tdWx0aTozNDIxOTIxNS02MzcyLTQ0NDQtOTViYi04MDJjZTViZTk4OWEiLCJlbmNyeXB0aW9uS2V5IjoiOXd5UTY5M1N6U1JKd3B3ZTg4bU1wYklsR05Ea3pLQ0tyd19TWnJFVGJ1MCIsImJyYW5jaCI6ImZlYXQvcm9jbS1hbWQtdXBzdHJlYW0iLCJyZXBvT3duZXIiOiJOb3ZhU2t5LUFJIiwicmVwb05hbWUiOiJTa3lSTCIsInByb3ZpZGVyIjoiZ2l0aHViIn19)
> 
> Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit [c0b63f4](https://github.com/NovaSky-AI/SkyRL/commit/c0b63f4ba6f3b0492c0e3cccb9a293e712891430). Configure [here](https://www.cursor.com/dashboard/bugbot).

This was pushed by mistake. Reverted back to https://github.com/amd-pratmish/SkyRL/commit/3ddb108cc85c969872c6f41f75810a72c4dfac03
